lib/streamlit/elements/vega_charts.py

```diff
@@ -12,85 +12,186 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-"""A Python wrapper around Altair.
-Altair is a Python visualization library based on Vega-Lite,
-a nice JSON schema for expressing graphs and charts.
-"""
+"""Collection of chart commands that are rendered via our vega-lite chart component."""
+
 from __future__ import annotations
 
+import json
 from contextlib import nullcontext
-from datetime import date
-from enum import Enum
-from typing import TYPE_CHECKING, Any, Collection, Literal, Sequence, cast
+from typing import TYPE_CHECKING, Any, Final, Literal, Sequence, cast
 
-import streamlit.elements.arrow_vega_lite as arrow_vega_lite
+import streamlit.elements.lib.dicttools as dicttools
 from streamlit import type_util
-from streamlit.color_util import (
-    Color,
-    is_color_like,
-    is_color_tuple_like,
-    is_hex_color_like,
-    to_css_color,
+from streamlit.elements.lib.built_in_chart_utils import (
+    AddRowsMetadata,
+    ChartType,
+    generate_chart,
 )
-from streamlit.elements.altair_utils import AddRowsMetadata
-from streamlit.elements.arrow import Data
-from streamlit.elements.utils import last_index_for_melted_dataframes
-from streamlit.errors import Error, StreamlitAPIException
+from streamlit.errors import StreamlitAPIException
 from streamlit.proto.ArrowVegaLiteChart_pb2 import (
     ArrowVegaLiteChart as ArrowVegaLiteChartProto,
 )
 from streamlit.runtime.metrics_util import gather_metrics
 
 if TYPE_CHECKING:
     import altair as alt
-    import pandas as pd
 
+    from streamlit.color_util import Color
     from streamlit.delta_generator import DeltaGenerator
+    from streamlit.elements.arrow import Data
+
+# See https://vega.github.io/vega-lite/docs/encoding.html
+_CHANNELS: Final = {
+    "x",
+    "y",
+    "x2",
+    "y2",
+    "xError",
+    "xError2",
+    "yError",
+    "yError2",
+    "longitude",
+    "latitude",
+    "color",
+    "opacity",
+    "fillOpacity",
+    "strokeOpacity",
+    "strokeWidth",
+    "size",
+    "shape",
+    "text",
+    "tooltip",
+    "href",
+    "key",
+    "order",
+    "detail",
+    "facet",
+    "row",
+    "column",
+}
+
+
+def _prepare_vega_lite_spec(
+    spec: dict[str, Any] | None = None,
+    use_container_width: bool = False,
+    **kwargs,
+) -> dict[str, Any]:
+    # Support passing no spec arg, but filling it with kwargs.
+    # Example:
+    #   marshall(proto, baz='boz')
+    if spec is None:
+        spec = dict()
+    else:
+        # Clone the spec dict, since we may be mutating it.
+        spec = dict(spec)
+
+    # Support passing in kwargs. Example:
+    #   marshall(proto, {foo: 'bar'}, baz='boz')
+    if len(kwargs):
+        # Merge spec with unflattened kwargs, where kwargs take precedence.
+        # This only works for string keys, but kwarg keys are strings anyways.
+        spec = dict(spec, **dicttools.unflatten(kwargs, _CHANNELS))
```

Comment:
We seem to potentially do a double clone here of `spec` if `kwargs` is given. I think we could just write

```python
if spec is None:
  spec = dict()

if len(kwargs):
  # Merge spec with unflattened kwargs, where kwargs take precedence.
  # This only works for string keys, but kwarg keys are strings anyways.
  spec = dict(spec, **dicttools.unflatten(kwargs, _CHANNELS))

if len(spec) == 0:
  raise StreamlitAPIException("Vega-Lite charts require a non-empty spec dict.")
```

Reply:
The spec also gets mutated in other parts later. So I believe its safer to always clone it..

Reply:
Change it 👍